### DISCUSSION TITLE
Converting the project back to Ruby (JRuby) - Update: Nevermind, not gonna happen :)

### DISCUSSION THREAD

**Discussion by vitobotta** [vitobotta/hetzner-k3s#339] [Announcements]
I started working on the project again after a break today while also doing some planning, and I came to the realisation that working with Crystal is slowing me down tremendously. Like, too much. I am a Rubyist at heart, and things I can do in a snap with Ruby often take ages to figure out how to do them in Crystal due to the lack of documentation and examples online. Heck, even when I try asking ChatGPT4 or Claude 3 Opus for help, they don't know much about Crystal and waste even more of my time. Crystal is also a problem for many people who would like to contribute to the project but are not familiar with the language, which isn't very popular.

So, because I would like to be faster at implementing features and fixing issues, I decided to switch back to Ruby. Or, more specifically, JRuby. From the user point of view, the only downside is that you will have to have Java installed on your system and run a .jar instead of a standalone executable.

So it's decided. Tomorrow I will start a new branch with the objective of converting the project back to Ruby.

**Comment by ThomasVitale** [comment]
In case it might be useful, I'd like to mention that with GraalVM it would be possible to develop on the JVM (similar to JRuby), but the application could also be compiled into a native executable so to remove the need for end-users to have a JVM installed. 

GraalVM is a popular choice in the Java ecosystem to compile applications as native executable, with features like instant start-up time, no warm-up, and low resource consumption. I typically use GraalVM in Java for cloud native applications (especially in serverless architectures) and for CLI tools. Besides Java, GraalVM also supports other languages, including Ruby.

General info on GraalVM's Ruby support: https://www.graalvm.org/ruby/
Detailed documentation: https://www.graalvm.org/latest/reference-manual/ruby/

**Reply by vitobotta** [reply]
Hi @ThomasVitale, this sounds awesome, I hadn't thought of TruffleRuby. I am going to look into it as it may be a better option. Thanks for the suggestion!

**Reply by vitobotta** [reply]
@ThomasVitale I am having a hard time finding information on how to create standalone executables with TruffleRuby. Could you point me in the right direction?

**Reply by vitobotta** [reply]
After reading more about it, I couldn't find a confirmation that I can compile **my program** to a standalone executable. All references I found where about TruffleRuby itself being compiled to a standalone executable. For now I will continue with JRuby but please let me know if I misunderstood something.

**Comment by jonas-tm** [comment]
Someone already did some kind of a rewrite in Java [here](https://github.com/easystartup-io/k3s-simplified).
And also [this](https://github.com/janosmiko/hetzner-k3s) Go version.

**Reply by vitobotta** [reply]
Yeah I saw it posted on Reddit and I didn't like the idea. He could have contributed to this project instead of cloning it. 

As for my rewrite, I managed to get working something I was having some difficulties with Crystal, so I may not be rewriting it in Ruby for now.
